src/shogun/mathematics/linalg/internal/implementation/VectorMean.h

```diff
@@ -0,0 +1,134 @@
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * The views and conclusions contained in the software and documentation are those
+ * of the authors and should not be interpreted as representing official policies,
+ * either expressed or implied, of the Shogun Development Team.
+ */
+
+#ifndef VECTOR_MEAN_IMPL_H_
+#define VECTOR_MEAN_IMPL_H_
+
+#include <shogun/lib/config.h>
+#include <shogun/lib/SGVector.h>
+#include <shogun/mathematics/linalg/internal/implementation/VectorSum.h>
+
+#ifdef HAVE_VIENNACL
+#include <shogun/lib/GPUVector.h>
+#include <shogun/lib/GPUMatrix.h>
+#endif
+
+#include <shogun/mathematics/eigen3.h>
+
+#include <numeric>
+
+namespace shogun
+{
+
+namespace linalg
+{
+
+/**
+ * All backend specific implementations are defined within this namespace
+ */
+namespace implementation
+{
+
+/**
+ * @brief Generic class vector_mean which provides a static compute method. This class
+ * is specialized for different types of vectors and backend, providing a mean
+ * to deal with various vectors directly without having to convert
+ */
+template <enum Backend, class Vector>
+struct vector_mean
+{
+	/** Scalar type */
+	typedef typename Vector::Scalar T;
+
+	/**
+	 * Method that computes the vector mean
+	 *
+	 * @param a vector whose mean has to be computed
+	 * @return the vector mean \f$\mean_i a_i\f$
+	 */
+	static T compute(Vector a);
+};
+
+
+/**
+ * @brief Specialization of generic vector_mean for the Eigen3 backend
+ */
+template <class Vector>
+struct vector_mean<Backend::EIGEN3, Vector>
+{
+	/** Scalar type */
+	typedef typename Vector::Scalar T;
+
+	/**
+	 * Method that computes the mean of SGVectors using Eigen3
+	 *
+	 * @param vec a vector whose mean has to be computed
+	 * @return the vector mean \f$\mean_i a_i\f$
+	 */
+	static T compute(SGVector<T> vec)
+	{
+		REQUIRE(vec.vlen > 0, "Vector can not be empty!\n");
+                return (vector_sum<Backend::EIGEN3, SGVector<T> >::compute(vec)
```

Comment:
Hi @OXPHOS. Thanks for sharing your thoughts.
- Actually I think the return type should be whatever eigen `mean` returns for this specialization. But it should never be just `T`. If `T` is of any integer type, it give wrong results. Do you think it would be doable using any of C++11 magic? I am talking about that 
```c++
auto foo() -> decltype(...)
```
thing. Something like [this](https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/mathematics/linalg/internal/modules/ElementwiseOperations.h#L56)
- You're right, there are just too many of them. Although all of them serve different purposes. But yes we'd like to clean things up a bit, hopefully, in this year's GSoC :)
About handling `SGVector` and `SGMatrix`, I'd say that we can just create an `Eigen::Map` of matrix type (the vector is just a matrix with 1 column) and then call `mean()` on that method. All we need to achieve this is to specify the data pointer, number of rows and cols. The method then can take a template instead. Kinda like [this one here](https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/mathematics/linalg/internal/implementation/ElementwiseUnaryOperation.h#L95).

The reason I am excited to have just one method is that, while using the API, I won't need to specify `linalg::vector_mean(...)` or `linalg::matrix_mean(...)` each time. I'll just call `linalg::mean(...)` and expect that it knows how to handle this.